File: benchmark/python/todo.py
#!/usr/bin/env python3
# Todo backend — Python
# In-memory CRUD for a todo list
# Operations: create, read, update, delete, list, filter

from dataclasses import dataclass, field
from typing import Optional

@dataclass
class Todo:
    id:    int
    title: str
    done:  bool = False

todos:   list[Todo] = []
next_id: int = 1

def create(title: str) -> str:
    global next_id
    todo = Todo(id=next_id, title=title)
    todos.append(todo)
    next_id += 1
    return f"created {todo.id}"

def get_by_id(id: int) -> str:
    match = next((t for t in todos if t.id == id), None)
    if match:
        return f"{match.id}: {match.title} [{match.done}]"
    return f"not found: {id}"

def complete(id: int) -> str:
    for t in todos:
        if t.id == id:
            t.done = True
            return f"completed {id}"
    return f"not found: {id}"

def delete(id: int) -> str:
    global todos
    before = len(todos)
    todos = [t for t in todos if t.id != id]
    if len(todos) < before:
        return f"deleted {id}"
    return f"not found: {id}"

def list_all() -> str:
    for t in todos:
        print(f"{t.id}: {t.title} [{t.done}]")
    return f"listed {len(todos)} todos"

def list_pending() -> str:
    pending = [t for t in todos if not t.done]
    for t in pending:
        print(f"{t.id}: {t.title}")
    return f"{len(pending)} pending"

def count() -> str:
    total   = len(todos)
    done    = sum(1 for t in todos if t.done)
    pending = total - done
    return f"total:{total} done:{done} pending:{pending}"

# --- run demo ---

create("Buy groceries")
create("Write tests")
create("Deploy to production")
create("Review PR")
create("Update docs")

list_all()
complete(1)
complete(3)
list_pending()
delete(2)
count()
print(get_by_id(3))
print(get_by_id(99))

File: benchmark/go/todo.go
// Todo backend — Go
// In-memory CRUD for a todo list
// Operations: create, read, update, delete, list, filter

package main

import "fmt"

type Todo struct {
	ID    int
	Title string
	Done  bool
}

var todos  []Todo
var nextID = 1

func create(title string) string {
	todo := Todo{ID: nextID, Title: title, Done: false}
	todos = append(todos, todo)
	nextID++
	return fmt.Sprintf("created %d", todo.ID)
}

func getByID(id int) string {
	for _, t := range todos {
		if t.ID == id {
			return fmt.Sprintf("%d: %s [%v]", t.ID, t.Title, t.Done)
		}
	}
	return fmt.Sprintf("not found: %d", id)
}

func complete(id int) string {
	for i := range todos {
		if todos[i].ID == id {
			todos[i].Done = true
			return fmt.Sprintf("completed %d", id)
		}
	}
	return fmt.Sprintf("not found: %d", id)
}

func deleteTodo(id int) string {
	before := len(todos)
	filtered := todos[:0]
	for _, t := range todos {
		if t.ID != id {
			filtered = append(filtered, t)
		}
	}
	todos = filtered
	if len(todos) < before {
		return fmt.Sprintf("deleted %d", id)
	}
	return fmt.Sprintf("not found: %d", id)
}

func listAll() string {
	for _, t := range todos {
		fmt.Printf("%d: %s [%v]\n", t.ID, t.Title, t.Done)
	}
	return fmt.Sprintf("listed %d todos", len(todos))
}

func listPending() string {
	count := 0
	for _, t := range todos {
		if !t.Done {
			fmt.Printf("%d: %s\n", t.ID, t.Title)
			count++
		}
	}
	return fmt.Sprintf("%d pending", count)
}

func count() string {
	total, done := len(todos), 0
	for _, t := range todos {
		if t.Done {
			done++
		}
	}
	return fmt.Sprintf("total:%d done:%d pending:%d", total, done, total-done)
}

func main() {
	create("Buy groceries")
	create("Write tests")
	create("Deploy to production")
	create("Review PR")
	create("Update docs")

	listAll()
	complete(1)
	complete(3)
	listPending()
	deleteTodo(2)
	fmt.Println(count())
	fmt.Println(getByID(3))
	fmt.Println(getByID(99))
}
